src/verifier/verifier.rs

```diff
@@ -19,7 +23,215 @@ pub fn build_symbol_table(ast: &Whamm, err: &mut ErrorGen) -> SymbolTable {
     visitor.table
 }
 
+#[allow(dead_code)]
+struct TypeChecker {
+    table: SymbolTable,
+}
+
+
+impl WhammVisitor<Option<DataType>> for TypeChecker {
+    fn visit_whamm(&mut self, whamm: &Whamm) -> Option<DataType> {
+
+        let mut err = ErrorGen::new("".to_string(), "".to_string(), 0);
+        self.table = build_symbol_table(whamm, &mut err);
+
+        // not printing events and globals now
+        for script in &whamm.scripts {
+            self.visit_script(script);
+        }
+
+        None
+    }
+
+    fn visit_script(&mut self, script: &Script) -> Option<DataType> {
+        for (_, provider) in &script.providers {
+            self.visit_provider(provider);
+        }
+        None
+    }
```

Comment:
Yep, looks like you don't call `enter_scope` and `exit_scope` on the symbol table as you traverse the AST. Try adding those on `visit_*` functions that would be entering some new scope (concentric circle) in the AST, it will fix your issue.

I just looked and the `verifier/builder_visitor.rs` has to enter/exit scopes as it's traversing the AST (this will actually add them to the symbol table as its being built. Feel free to trace through that visitor to see examples of what I mean by entering/exiting scopes above.